src/dos/drive_fat.cpp

```diff
@@ -1056,15 +1082,24 @@ bool fatDrive::FileExists(const char *name) {
 bool fatDrive::FileOpen(DOS_File **file, char *name, uint32_t flags) {
 	direntry fileEntry;
 	uint32_t dirClust, subEntry;
-	if(!getFileDirEntry(name, &fileEntry, &dirClust, &subEntry)) return false;
-	/* TODO: check for read-only flag and requested write access */
+
+	if (!getFileDirEntry(name, &fileEntry, &dirClust, &subEntry)) {
+		DOS_SetError(DOSERR_FILE_NOT_FOUND);
+		return false;
+	}
+
+	if ((readonly || (fileEntry.attrib & DOS_ATTR_READ_ONLY)) &&
+	    (flags & 0xf) != OPEN_READ) {
+		DOS_SetError(DOSERR_ACCESS_DENIED);
+		return false;
```

Comment:
There's a fair bit happening all in this `if` body.

 - the first `readonly` -> that's against the entire filesystem?

 - the second `fileEntry.attrib & DOS_ATTR_READ_ONLY` -> that's against this specific file attribute.
 - The third `(flags & 0xf) != OPEN_READ` is a bit less obvious. If any flags are used besides `OPEN_READ` then they get access denied? So even read and write would fail this test? 

Suggest lifting the latter two out out to self-documenting bool names to make the `if` read more like a sentence.

Reply:
Yeah? Pretty sure r/w open is supposed to fail if the target is r/o. It works like that on Windows.